ARCs/arc-draft_extended_desc.md

```diff
@@ -0,0 +1,160 @@
+author: Joe Polny (@joe-p)
+discussions-to: <URL>
+status: Draft
+type: Standards Track
+category: ARC
+created: 2023-11-14
+requires: 4
+---
+
+
+## Abstract
+This ARC takes the existing JSON description of a contract as described in ARC4 and adds more fields for the purpose of client interaction
+
+## Motivation
+The description provided by ARC4 is missing a lot of critical information that clients should know when interaction with an app.
+
+## Specification
+```ts
+/** Mapping of named structs to the ABI type of their fields */
+interface StructFields {
+  [name: string]: string | StructFields;
+}
+
+/** Describes a single key in app storage */
+interface StorageKey {
+  /** The type of the key. Can be ABI type or named struct */
+  keyType: string;
+  /** The type of the value. Can be ABI type or named struct */
+  valueType: string;
+  /** The key itself, as a byte array */
+  key: number[];
+}
+
+interface StorageMap {
+  /** The type of the key. Can be ABI type or named struct */
+  keyType: string;
+  /** The type of the value. Can be ABI type or named struct */
+  valueType: string;
+  /** The prefix of the map, as a string */
+  prefix: string;
+}
+
+interface Method {
+  /** The name of the method */
+  name: string;
+  /** Optional, user-friendly description for the method */
+  desc?: string;
+  /** The arguments of the method, in order */
+  args: Array<{
+    /** The type of the argument */
+    type: string;
+    /** If the type is a struct, the name of the struct */
+    struct?: string;
+    /** Optional, user-friendly name for the argument */
+    name?: string;
+    /** Optional, user-friendly description for the argument */
+    desc?: string;
+  }>;
+  /** Information about the method's return value */
+  returns: {
+    /** The type of the return value, or "void" to indicate no return value. */
+    type: string;
+    /** If the type is a struct, the name of the struct */
+    struct?: string;
+    /** Optional, user-friendly description for the return value */
+    desc?: string;
+  };
+  /** an action is a combination of call/create and an OnComplete */
+  actions: {
+    /** OnCompeltes this method allows when appID === 0 */
+    create: ('NoOp' | 'OptIn' | 'DeleteApplication')[];
+    /** OnCompeltes this method allows when appID !== 0 */
+    call: ('NoOp' | 'OptIn' | 'CloseOut' | 'ClearState' | 'UpdateApplication' | 'DeleteApplication')[];
+  };
+  readonly: boolean;
+}
+
+interface Contract {
+  /** A user-friendly name for the contract */
+  name: string;
+  /** Optional, user-friendly description for the interface */
+  desc?: string;
+  /**
+   * Optional object listing the contract instances across different networks
+   */
+  networks?: {
+    /**
+     * The key is the base64 genesis hash of the network, and the value contains
+     * information about the deployed contract in the network indicated by the
+     * key
+     */
+    [network: string]: {
```

Comment:
```suggestion
    [network: 'mainnet' | 'testnet' | 'betanet' | string]: {
```

Reply:
Current one is base64 genesis hash, but I feel like we should support the more understandable short-hand?

Side-note: just realised this is part of ARC-0004. I've never seen anyone use this, it's a bit tricky if you are auto-generating the json file to include stateful data like this. We also don't currently support it in algokit utils, although if we want to support it we should add that! This would be handy when third parties like oracles are distributing a file.

Reply:
Part of the rationale for the base64 is that is allows one to accurately describe a private network. We could have a separate entry for public networks though, but because this is so seldomly used I'm not sure if it'd really be worth it.

Reply:
My suggestion above still has `string` so the implementation would be it's either base64 or one of the special known network names. It would make it a lot easier to use this feature.

Reply:
This is minor though happy if it's left out.